### PR TITLE
add in pycbc_multiifo_exclude_zerolag executable

### PR DESCRIPTION
**Pull request by GarethDaviesGW**
Adding in an executable to remove zerolag events from _all_ coincidence types in the background_exc coincidences

This means that the single-detector triggers which form a zerolag coincidence cannot be used in formation of a timeslide coincidence

Also a fix to the .remove() and .select() method on DictArrays, which was not acting on datasets within groups

### CODE REVIEW

**Comment by GarethDaviesGW** [comment]
Executable testing:
Exclude_foreground tested on O2 a21 `coinc_statmap` files
image
We see that the higher tails of the distribution, caused by signals in zerolag showing up in backgrounds where they dont form a coinc, are removed in the fg_exc data

Workflow creation testing:
The workflow creation script runs to completion, and in the main.dax file created, it creates the `EXCLUDE_FOREGROUND` files, and uses these as the later input into `coinc_statmap_inj`, `combine_statmap`, the `--background-files` input of `combine_statmap` when it handles injections,`plot-snrifar`, and `plot_snrratehist`.

POSSIBLE ISSUE/QUESTION: This plotting inputs are a possible problem, as after the exclusive background is reduced, then the ifar should be recalculated. should this be added in as well? It will get recalculated again by `add_statmap` if that is used

Further testing:
If needed, I can set up a run which reuses all files up to the `coinc_statmap`

File: bin/hdfcoinc/pycbc_multiifo_exclude_zerolag [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+"""
+Remove all coincs in background_exc that contain triggers at time of zerolag
+coincidences from *any* coincidence type with ifar above a certain threshold
+"""
+
+import h5py, numpy as np, argparse, logging, pycbc, pycbc.events, pycbc.io
+import pycbc.version
+import pycbc.conversions as conv
+from pycbc.events import veto, coinc
+from ligo import segments
```
Comment (tdent) [inline bin/hdfcoinc/pycbc_multiifo_exclude_zerolag]: can you remove unneeded imports?  I think pycbc.events and pycbc.conversions are not used at least.

File: pycbc/workflow/coincidence.py [hunk 2]
```diff
@@ -271,19 +271,31 @@ class PyCBCMultiifoCombineStatmap(PyCBCCombineStatmap):
...
     def create_node(self, statmap_files, background_files, tags=None):
+        node = super(PyCBCMultiifoAddStatmap, self).create_node(statmap_files,
+                                                            tags=tags)
+        if 'injections' in tags:
+            node.add_input_list_opt('--background-files', background_files)
+
+        return node
+
+class PyCBCMultiifoExcludeZerolag(Executable):
+    """ Remove times of zerolag coincidences of all types of from
```
Comment (tdent) [inline pycbc/workflow/coincidence.py]: typo -> "of all types from"

File: pycbc/workflow/coincidence.py [hunk 3]
```diff
@@ -262,7 +262,7 @@ def create_node(self, statmap_files, tags=None):
         node = Node(self)
         node.add_input_list_opt('--statmap-files', statmap_files)
         node.new_output_file_opt(statmap_files[0].segment, '.hdf',
-                                 '--output-file', tags=tags)
+                                 '--output-file', tags=None)
```
Comment (tdent) [inline pycbc/workflow/coincidence.py]: can you explain why this change to tags=None is needed / desirable

File: pycbc/io/hdf.py [hunk 4]
```diff
@@ -141,6 +142,7 @@ def __init__(self, data=None, files=None, groups=None):
 
         for k in self.data:
             setattr(self, k, self.data[k])
+        self.groups=groups
```
Comment (tdent) [inline pycbc/io/hdf.py]: could this be done right at the beginning of the init method just after self.data = data?  atm it's not totally clear if 'groups' might be changed by the block of code in between (I don't think it is)

File: pycbc/io/hdf.py [hunk 5]
```diff
@@ -964,19 +966,22 @@ def combine_and_copy(f, files, group):
...
-    for sk in fi[k].keys():
-        path = k + '/' + sk
-        if path not in dset:
-            if isinstance(fi[path], h5py.Dataset):
-                subkey_list.append(path.lstrip('/'))
-            else:
-                subkey_list += get_all_subkeys(fi, path, dset)
+    subkey_start = key
+    if key == '': grpk = grp
+    else: grpk = grp[key]
```
Comment (tdent) [inline pycbc/io/hdf.py]: will need to put in linebreaks/indentation here

**Comment by tdent** [comment]
Have asked Gareth to set up a test on a chunk21 run to test that the outputs are what we expect.

**Comment by tdent** [comment]
@GarethDaviesGW can you point to what was going wrong in the remove() and select() methods, if it's not super complicated?  the change to io.hdf seems a bit nontrivial - which is not necessarily bad, but means that it's doesn't look like a very simple fix.

File: pycbc/workflow/coincidence.py (hunk 3 above)
Comment (GarethDaviesGW) [reply pycbc/workflow/coincidence.py]: To be honest I don't think this is needed and will revert in this case

I did this because changing from `tags=tags` to `None` within the `PyCBCMultiifoExcludeZerolag.create_node()` meant that the tags didn't appear twice in the output filename, and thought this could be an issue elsewhere. It doesn't seem to actually make a difference in this case

**Comment by GarethDaviesGW** [comment]
The remove() and select() methods were not acting on the `ifo/something` datasets, this is as they were not recursively going through the ifo Group

By changing to loop through groups (and storing groups as part of self in order to do so), we ensure that the datasets which are stored within groups are select()ed or remove()d

File: pycbc/workflow/coincidence.py (hunk 3 above)
Comment (spxiwh) [reply pycbc/workflow/coincidence.py]: If the tags are appearing twice in the output filename then it is a problem. I don't think the issue is in this line though. Indeed removing this will just cause the `tags` kwarg to do nothing. I think the issue is that if you have already sent the tags when creating the Executable you should not send the same tags again when creating the Node.

File: pycbc/workflow/coincidence.py [hunk 6]
```diff
@@ -779,14 +818,35 @@ def setup_multiifo_combine_statmap(workflow, final_bg_file_list, bg_file_list,
...
+    Exclude single triggers close to zerolag triggers from forming any
+    background events
+    """
+    if tags is None:
+        tags = []
+    make_analysis_dir(out_dir)
+    logging.info('Setting up multiifo exclude zerolag')
+
+    exc_zerolag_exe = PyCBCMultiifoExcludeZerolag(workflow.cp, 'exclude_zerolag',
+                                                  ifos=workflow.ifos, tags=tags,
```
Comment (spxiwh) [inline pycbc/workflow/coincidence.py]: Yeah, look, you've sent the tags here when initializing the Executable.

File: pycbc/workflow/coincidence.py [hunk 7]
```diff
@@ -779,14 +818,35 @@ def setup_multiifo_combine_statmap(workflow, final_bg_file_list, bg_file_list,
...
+                                                  out_dir=out_dir)
+    exc_zerolag_node = exc_zerolag_exe.create_node(statmap_file,
+                                                   other_statmap_files,
+                                                   tags=tags)
```
Comment (spxiwh) [inline pycbc/workflow/coincidence.py]: So *don't* send them again here! This would be for tags that vary on the node level (which in this case is none of them).

Comment (GarethDaviesGW) [reply pycbc/workflow/coincidence.py]: Thanks for that, it doesn't repeat the tags which is all I was wanting.

One thing I am a bit confused by though is the naming of the file: the `exclude_foreground` files all start with H1L1V1, even when its a 2-detector output. However the `multiifo_statmap` ones all only have their used detectors. Is this something easy to change or even needed?

Examples:
`H1L1V1-EXCLUDE_ZEROLAG_FULL_2DET_H1V1-1185937218-687600.hdf`
`L1H1-MULTIIFO_STATMAP_FULL_DATA_FULL_2DET_H1L1_VETOES-1185937218-687600.hdf`

Comment (spxiwh) [reply pycbc/workflow/coincidence.py]: You can choose that. When initializing the executable you choose the set of ifos that are active. In many cases this is `workflow.ifos` in some cases this is a specific subset.

**Comment by tdent** [comment]
To note down a couple of simplifications here : We can actually just use the DictArray() class directly with `data=` for removing triggers from the exclusive background.  Then, some of the changes to the class are no longer needed or wanted. 
In any case the changes to `name_all_datasets`, though they end up not being required for the rest of the PR, simplify the code and do not break anything.

File: pycbc/workflow/coincidence.py (hunk 7 above)
Comment (GarethDaviesGW) [reply pycbc/workflow/coincidence.py]: Thanks! This is in now

**Comment by GarethDaviesGW** [comment]
> The remove() and select() methods were not acting on the `ifo/something` datasets, this is as they were not recursively going through the ifo Group
> 
> By changing to loop through groups (and storing groups as part of self in order to do so), we ensure that the datasets which are stored within groups are select()ed or remove()d

Note that this has now been reverted and uses DictArray directly

**Comment by tdent** [comment]
Can you briefly say what the change to workflow plotting is for?

**Comment by GarethDaviesGW** [comment]
> Can you briefly say what the change to workflow plotting is for?

Yes, this is to ensure that the snrchi job files are properly named they had been getting names e.g. `plot_snrchi-_...` due to empy strings being passed as tags. This could and maybe should be a separate PR, so I can take it out if requested

**Comment by tdent** [comment]
well, as usual, unrelated commits should not go in the same PR, is this change to plot snrchi tested? if so it can stay in, if not please remove from this.

**Comment by GarethDaviesGW** [comment]
> well, as usual, unrelated commits should not go in the same PR, is this change to plot snrchi tested? if so it can stay in, if not please remove from this.

It is tested, and produces sensibly named files
